liblangutil/SourceReferenceFormatterHuman.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author Christian <c@ethdev.com>
+ * @date 2014
+ * Formatting functions for errors referencing positions and locations in the source.
+ */
+
+#include <liblangutil/SourceReferenceFormatterHuman.h>
+#include <liblangutil/Scanner.h>
+#include <liblangutil/Exceptions.h>
+#include <libdevcore/termcolor.h>
+#include <cmath>
+#include <iomanip>
+
+using namespace std;
+using namespace dev;
+using namespace langutil;
+
+namespace
+{
+inline ostream& frameColor(ostream& os) { return os << termcolor::bold << termcolor::cyan; };
+inline ostream& messageColor(ostream& os) { return os << termcolor::bold << termcolor::white; };
+inline ostream& errorColor(ostream& os) { return os << termcolor::bold << termcolor::red; };
+inline ostream& highlightColor(ostream& os) { return os << termcolor::yellow; };
+inline ostream& diagColor(ostream& os) { return os << termcolor::bold << termcolor::yellow; };
+}
+
+void SourceReferenceFormatterHuman::printSourceLocation(SourceReference const& _ref)
+{
+	if (_ref.position.line < 0)
+		return; // Nothing we can print here
+
+	int const leftpad = static_cast<int>(log10(max(_ref.position.line, 1))) + 1;
+
+	// line 0: source name
+	m_stream << string(leftpad, ' ') << frameColor << "--> " << termcolor::reset
+			 << _ref.sourceName << ":" << (_ref.position.line + 1) << ":" << (_ref.position.column + 1) << ": "
+			 << '\n';
+
+	if (!_ref.multiline)
+	{
+		int const locationLength = _ref.endColumn - _ref.startColumn;
+
+		// line 1:
+		m_stream << string(leftpad, ' ') << frameColor << " |" << termcolor::reset << '\n';
```

Comment:
Why would you print an empty line? To have spacing and make the source snippet stand out?

Reply:
For the diagnostics output format I got literally inspired by the output format of the Rust compiler (just like the original PR from the external contribution).

Apart from that, I feel that more space around makes the actual text also more readable.